consensus/src/metrics_safety_rules.rs

```diff
@@ -32,7 +32,9 @@ impl MetricsSafetyRules {
         let proofs = self
             .storage
             .retrieve_epoch_change_proof(sr_waypoint.version())
-            .expect("Unable to retrieve Waypoint state from Storage");
+            .map_err(|_| {
+                Error::InternalError("Unable to retrieve Waypoint state from storage".into())
+            })?;
```

Comment:
this is interesting, do we want consensus (libra-node) to crash if it cannot sync LSR -- cc: @zekun000

Reply:
There are several considerations here and I'd like to weigh in on some of them. First of all `expect`. This is not used consistently, which makes it hard to read. Much of what I've seen to date, suggests that people use it as "it is a programming mistake if this happens, but I pretty much guarantee that it wont". It can also mean "this might happen, but if it does things are hosed and the best we can do is to crash here".

The proposed change introduces yet another meaning: "This might happen and if it does, the caller can decide what to do about it." (This is not unreasonable, if an invariant is found to be violated, a program can carry on if there is a reasonable way to repair the program state so that the invariant is restored. Perhaps the caller can do this, but in this case I think David is saying, he is not so sure.)

I don't like it when things are ambiguous like this. It makes the code hard to read and it makes it difficult for an outsider to flag this as a potential error since it refers to considerations that are either undocumented, or documented elsewhere in a way that requires a substantial up front cost before you can properly read the code.

My recommendation is the make things as explicit as possible. 

- If it is the case that the caller can conceivably repair the program state, then return the error.
- If this can only happen in a situation where the process should stop immediately, then do `.unwrap_or_else(|| unrecoverable!(".."))`
- If this should not happen at all (i.e. it is a programming mistake if it happens) then explicitly do a `checked_assume!(proofs.is_some()); // justification for the assumption`

Reply:
@fga, I think the easy way to support doing this is to state that the only use of perform_initialize itself has an expect to resolve the result and therefor this ehis expect is duplicitous. At the same time, however, you are losing the internal error string, why was it unable, was that message meaningful. We are losing some context.

Reply:
If you want to maintain the behavior of expect, then terminate the execution in this context by all means. I would recommend, however, that instead of expect, you either unwrap_or_else or do a checked_assume. That makes it clear whether you expect the unwrap to succeed or whether you expect to to sometimes fail. Making this distinction clear to the reader (which most likely is your future self) will be a great help to them.

Reply:
I'm not sure I follow. I would think unwrap is maybe unclear, but an expect statement that doesn't state why this is expected is a fail. At the same time, when using locks, folks use unwrap religiously. It seems a shame to add a lot of boiler plate code where ideally where we already have unwrap / expect :/